src/garbage-collection.md

```diff
@@ -0,0 +1,105 @@
+---
+description: Everything about garbage collection in AssemblyScript.
+sidebarDepth: 0
+---
+
+# Garbage Collection
+
+::: tip NOTE
+This section covers the new runtime added in AssemblyScript 0.18. The previous runtime documentation [can be found here](./runtime.md).
+:::
+
+While the WebAssembly [GC](https://github.com/WebAssembly/gc) 🦄 proposal is still in the works, AssemblyScript ships with multiple garbage collector runtimes implemented on top of linear memory that are useful for different use cases. Their interface is always the same, so these are interchangeable, yet differ in their level of sophistication.
+
+## Runtime variants
+
+With the `--runtime` compiler option, one of multiple runtime variantes can be selected, which determines the memory allocator and garbage collector included in a program:
+
+```
+  --runtime             Specifies the runtime variant to include in the program.
+
+                         default      TLSF + lightweight GC invoked externally
+                         incremental  TLSF + incremental GC (experimental)
+                         stub         Minimal runtime stub (never frees)
+                         ...          Path to a custom runtime implementation
+```
+
+Using the `--exportRuntime` compiler option, the runtime interface can be exported to the host to enable external allocation etc.
+
+```
+  --exportRuntime       Exports the runtime helpers (__new, __collect etc.).
+```
+
+To interface with the runtime externally, you'll probably want to use the [loader](./loader.md) as well.
+
+## "Default" runtime
+
+The default runtime provides a complete dynamic memory manager backing `new`, `heap.alloc`, `heap.realloc` and `heap.free` and is suitable for most non-interactive use cases, like the compiler itself.
+
+It aims at not bloating generated modules, i.e. as would be the case by going overboard with filler implementations for WebAssembly futures that are not yet available (e.g. maintaining a separate shadow stack since WebAssembly's execution stack is not inspectable). As a compromise, it has to be invoked manually (it is not automated), ideally externally, by calling `exports.__collect()` when the WebAssembly execution stack is known to be fully unwound.
+
+It always performs a full garbage collection cycle, marking all reachable and sweeping all unreachable objects, and stops the program while doing so. Invoking garbage collection more often reduces the amount of garbage lingering around at any point in time, while invoking it less frequently increases throughput.
+
+**Usage instructions:**
```

Comment:
I think these usage instructions on each runtime mode are great! I think what would be helpful, is also maybe providing a commented example of these use cases (Not multiple, maybe just one), where you explain what would happen.

For example, without ever using the runtime :joy: 


```typescript
let myObjPtr = myWasm.getPtr();
// We will pin this, so that while were are in Javascript (Our Wasm host) we won't lose the object
myWasm.__pin(myObjPtr);

for (let i = 0; i < 100; i++) {
  // Collect all the garbage, but pinned objects will not be removed, because we pinned them
  myWasm.__collect();
}

// ...
```

You get what I mean haha :joy: I'd also think it'd maybe to show what's going on in the Wasm as well, and why :smile: As personally, I'm honestly still a little confused on how this would work from an all Wasm (No objects ever need to be used by the host, I just throw things in a buffer in linear memory) or Wasi perspective :smile:

Reply:
Added a few simple examples for each runtime, but didn't bother going overboard with `__pin`. Just showing typical loops of usage.